### PR TITLE
Fix iterating over entries of parallel PETSc matrices.

### PR DESCRIPTION
**Pull request by bangerth**
Previously, it was not possible to iterate over the local range because
one would have to call
  matrix.end(row)
where 'row' is the last locally owned row, and that triggered an
assertion because this end iterator is also the begin iterator of
the next row -- which is not locally owned any more.

Fix this.

### CODE REVIEW

**Comment by drwells** [comment]
I'll look at this but I may not be able to do so for a day or two.

**Comment by bangerth** [comment]
No rush -- I have nothing else that depends/builds on this.

File: source/lac/petsc_matrix_base.cc [hunk 1]
```diff
@@ -34,11 +34,14 @@ namespace PETScWrappers
...
     {
-      // if we are asked to visit the
-      // past-the-end line, then simply
-      // release all our caches and go on
-      // with life
-      if (this->a_row == matrix->m())
+      // if we are asked to visit the past-the-end line (or a line that is not
+      // stored on the current processor), then simply release all our caches
+      // and go on with life
+      if ((this->a_row == matrix->m())
```
Comment (drwells) [inline source/lac/petsc_matrix_base.cc]: This condition should now be redundant, right? No matrix is going to store the `m`th row.

File: tests/petsc/iterate_parallel_01.cc [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+// check re-initializing a preconditioner (parallel version)
```
Comment (drwells) [inline tests/petsc/iterate_parallel_01.cc]: Describe the original bug here, please.

File: tests/petsc/iterate_parallel_01.cc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    pre.vmult(dst, src);
+    dst.print(deallog.get_file_stream());
+
+    mat.add(0,0,1.0);
+    mat.compress(VectorOperation::add);
+
+    pre.initialize(mat);
+    pre.vmult(dst, src);
+    dst.print(deallog.get_file_stream());
+  }
```
Comment (drwells) [inline tests/petsc/iterate_parallel_01.cc]: This part doesn't seem relevant to the rest of the test (for iterators). Can we remove this block?

File: tests/petsc/iterate_parallel_01.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  //////////////////////////////////////////////
+  /////This is a test for the local matrix iterator
+  //////////////////////////////////////////////
+  unsigned int start_row = mat.local_range().first;
+  unsigned int end_row = mat.local_range().second;
+  for (auto r = start_row; r < end_row; ++r)
+    {
+      for (auto itr = mat.begin(r); itr != mat.end(r); ++itr)
+        {
+          //do nothing, just iterate
```
Comment (drwells) [inline tests/petsc/iterate_parallel_01.cc]: We should probably print something in case the compiler optimizes the loop away.

File: tests/petsc/iterate_parallel_01.cc [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+{
+  Utilities::MPI::MPI_InitFinalize mpi_initialization (argc, argv, 1);
+  MPILogInitAll log;
+
+  test<PETScWrappers::PreconditionJacobi> ();
+  test<PETScWrappers::PreconditionBlockJacobi> ();
+  test<PETScWrappers::PreconditionEisenstat> ();
+  test<PETScWrappers::PreconditionBoomerAMG> ();
+  test<PETScWrappers::PreconditionParaSails> ();
+  test<PETScWrappers::PreconditionNone> ();
```
Comment (drwells) [inline tests/petsc/iterate_parallel_01.cc]: Same issue: we don't need to run the test over multiple preconditioners (their state does not impact the iterator).

File: doc/news/changes/minor/20180320Bangerth-2 [hunk 6]
```diff
@@ -0,0 +1,9 @@
+Fixed: Previously, it was not possible to iterate over the local range
+of rows of PETSc matrix objects because one would have to call
+`matrix.end(row)` where `row` is the last locally owned row, and that
```
[the remark this answers is not in the document]
Comment (bangerth) [reply doc/news/changes/minor/20180320Bangerth-2]: No, these are actually run through doxygen.

File: include/deal.II/lac/petsc_matrix_base.h [hunk 7]
```diff
@@ -1504,13 +1509,23 @@ namespace PETScWrappers
...
+    // in the parallel case, we need to put it on the first entry of
+    // the first row after the locally owned range. this of course
+    // doesn't exist, but we can nevertheless create such an
+    // iterator. we need to check whether 'i' is past the locally
+    // owned range of rows first, before we ask for the length of the
+    // row since the latter query leads to an exception in case we ask
+    // for a row that is not locally owned
     for (size_type i=r+1; i<m(); ++i)
-      if (row_length(i) > 0)
+      if (i == local_range().second || (row_length(i) > 0))
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/lac/petsc_matrix_base.h]: Correct. Think of cases such as a block decomposition of the Stokes operator -- then the bottom left block will have empty rows for pressure DoFs that are constrained by hanging nodes. We have similar code for all other sparse matrix iterators.

File: source/lac/petsc_matrix_base.cc (hunk 1 above)
Comment (bangerth) [reply source/lac/petsc_matrix_base.cc]: I suspect that to be true -- I had left it in because I wasn't quite sure whether `local_range.second` is always at most equal to `m()`. But it must be, and if that's not the case, then there's clearly a bug here. So I'll rewrite this condition.

**Comment by bangerth** [comment]
I've updated the PR per your comments -- thanks for taking a look!

(I had simply copied the test verbatim from the one by Feimi Yu, without any adaptation. That may have been a bit too lazy ;-) )

File: include/deal.II/lac/petsc_matrix_base.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/lac/petsc_matrix_base.h]: Good point.

File: tests/petsc/iterate_parallel_01.cc [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+void test ()
+{
+  unsigned int myid = Utilities::MPI::this_mpi_process (MPI_COMM_WORLD);
+  unsigned int numproc = Utilities::MPI::n_mpi_processes (MPI_COMM_WORLD);
+
+  deallog << numproc << std::endl;
+
+  // each processor owns 2 indices and all
+  // are ghosting Element 1 (the second)
```
Comment (drwells) [inline tests/petsc/iterate_parallel_01.cc]: The code below has DoF 0 ghosted on processor 1, but this says DoF 1.

**Comment by drwells** [comment]
Looks good aside from the (irrelevant to the test, as far as I know) erroneous comment.

**Comment by bangerth** [comment]
Patch updated as requested.

**Comment by drwells** [comment]
It looks like Travis is not happy due to a misaligned `<<`; we should be done once that is fixed.

**Comment by bangerth** [comment]
Good grief. This should now be fixed. Sorry for the many round-trips.
